### ISSUE TITLE
Improve the rendering of RTL constants

### ISSUE THREAD

**Issue by uxmal** [uxmal/reko#852]
There are a few problems around the rendering of Reko's RTL `Constant` numbers. I'm writing this proposal to illustrate the problems, propose a solution, and solicit feedback/improvements.

Numbers are rendered unnecessarily verbosely. For instance, Reko writes a 64-bit zero as `0x000000000000000`. The reason why this is done is to make it clear to the reader that this is a 64-bit number, and not a 16-bit number like `0x0000`. However, when writing unit tests it's very tedious and error-prone to type those long sequences of `0`s.

Ah, but did you notice that there actually were only 15 `0`'s in the first example above? Should it then be considered a 60-bit number? What is the type of that constant, perhaps `word60`? At the RTL stage Reko needs to maintain an accurate representation the sizes of expressions. Relying on the length of the text representation is error prone, and doesn't allow the representation of bit vectors whose length isn't an exact multiple of 4 (consider the 18-bit constants on PDP-10 computers)

Another concern is being able to save RTL in a persistent format and be able to read it back faithfully. There is no way to tell if `0x0000` is a signed or unsigned integer, or just a 16-bit word. 

### Word and integer constants
This proposal introduces a suffix to the textual representation numbers in Reko RTL. This is analogous to the `F` sigil used for single precision floating point numbers in C or C# (e.g. `3.1412F`). The suffix should indicate the bit size of the number, and whether or not it is a typeless word or signed/unsigned integer. This proposal uses a `_` sigil for typeless words, an `i` sigil for signed integers, and a `u` sigil for unsigned integers. Some examples:
```
// Typeless words
         0_32
         0_64       // not the same as 0_32 or 0_16
0x00123400_32
  0x123400_32       // Same 32-bit word, with no leading 0's
       0x42_8
  0o123456_18       // 18-bit quantity rendered in octal (support for
                    // octal numbers is not part of the scope of the
                    // proposal)

// Signed integers
         0i32
         1i32
        -9i64

// Unsigned integers
    0x1234u16
      0x34u16

```

### Taking advantage of the word and pointer size
The 3-character suffixes can be made shorted by taking advantage of the specific pointer and word sizes of computer architectures. These sizes would make it unnecessary to specify the sizes of constants of those types. On a 32-bit machine, a 32-bit typeless word would be represented simply as `0x1234`, while `0x1234i` and `0x1234u` would represent signed and unsigned 32-bit integers, respectively. For such a machine, a 16-bit word would still be displayed as `0x1234_16`.

The reason why typeless words are represented with no suffix is that the majority of RTL constants generated by Reko during disassembly will be typeless until type analysis has assigned types to them (see 'High-level language output' below).

### Pointers
Today, Reko supports two kinds of pointer typed constants -- also known as `Address`es. "Flat" addresses are rendered as `0x00123400`. To avoid confusing pointer typed constants with other types, a `p` sigil is used. The bit size can be omitted if the pointer size is the same as that of the current architecture.
```
// Pointers (addresses)
0x00123400p32       // 32-bit pointer
    0x123400p       // omit the bit size if the pointer size matches 
                    // that of the architecture
```
Segmented addresses are rendered today as `0800:1234`. A possible new rendering could be `0x800:1234p`. An explicit size could be used to represent segmented pointers on x86 32- or 64-bit systems (e.g. `0x0C00:42p48` reprenents a far pointer on a 32-bit x86 system).

### High-level language output
Note that this proposal doesn't affect the output high-level language (HLL). By the time HLL is generated, all pointer constants in the program should have been replaced with symbols like `ptr00123400` or `pdw0800_1234`. Other numerical contants are rendered as the the syntactic rules of the target HLL language dictate.

### Impact
Adopting this proposal would require updating all the Reko unit tests and regression tests, wherever RTL is being generated. This is a big work item -- there are thousands of unit tests -- but it would only happen once.

**Comment by ptomin** [comment]
Sounds reasonable. I agree that `0_64` looks better than `0x000000000000000`

**Comment by uxmal** [comment]
Please review the outputs of commit 3c8321b97f8f7a0a23b522d41394fd90d7bc1de3. I decided to use a different suffix scheme, to be consistent: `_iNN`, `_uNN` and `_pNN`. The consistent use of `_` is coincidentally consistent with the C++ 11 literal suffix syntax. There are a few bugs when generating `*.globals.c` files but I want to focus on the `.dis` files first. 

What do you think?

**Comment by gbody** [comment]
Using https://github.com/uxmal/reko/commit/3c8321b97f8f7a0a23b522d41394fd90d7bc1de3 and having looked at the output for M68K using the sample program LinkIIgs. Due to inconsistant mapping of constants, I initially found it a little confusing after reading the start and end sections of procedures where link and unlk instrustructions incremented/decremented a7 by 4 and the offset used for the link not having been defined with a _ sigil. There is RTL for other instructions that also didn't get defined. 
As to should the native size be excluded, I lean towards defining for all values (troubleshooting purposes). Why, as you have consistency and you can clearly see that the values are being interpreted as expected. 

If your using NN for the bit size, for the case of a byte or single bit Booleen, is that going to be defined with a leading zero. ie. byte 08 or single digit of 8.
I currently see it's single digit.

**Comment by uxmal** [comment]
@gbody: I've added a preprocessor switch `ALWAYS_SHOW_SIGIL` in commit 373421bc03607e3d2ea31842d3e459d994fbe9c8 which always emits the sigil, regardless of size, and also pads it out to at least 2 digits (so a byte constant will be rendered as `0x42_08` and a signed byte like `66_i08`. 

I've also added another preprocessor switch `USE_ANGR_STYLE_SIGILS` which renders bitvector constants a bit like the https://github.com/angr/angr project does. I think their IR format is called VEX. If you set this preprocessor switch, reko will render constants like this: 
* `0<64>` for a 64-bit word
* `123<i32>` for a 32-bit signed integer
* `0x1234<u16>` for a 16-bit unsigned integer

I find the angr-style format a little easier to read, but it complicates parsing a little bit. We don't want to confuse `0 < 64 >` with `0<64>`. However, the new format requires the sigil after all numeric constants, so finding `0 < 64 >` in the input stream is a syntax error.

Please try these two new alternatives and let me know what you think.

**Comment by ptomin** [comment]
I have tested 373421b.
I prefer `angr` style for constants. Using `_` as separator makes constants similar to SSA ids

**Comment by uxmal** [comment]
@ptomin: that's a good point. It's definitely easier to distinguish identifiers from constants visuall with `<32>` suffixes. Unless I hear loud complaints, then, I will proceed with the `angr`-style constants.

**Comment by gbody** [comment]
I have tested https://github.com/uxmal/reko/commit/373421bc03607e3d2ea31842d3e459d994fbe9c8, originally I found the angr sytle blended into some sections of the output, but after a bit of use it's fine. I am happy with the output and which ever style you decide on will be ok.

**Comment by uxmal** [comment]
Commit 5e6b8e4f6150be7ebbd876d462ecabc10a831be0 implements this proposal, using `angr`-style sigils. Now the reko IL format is round-trippable. An added benefit is that displaying the explicit bit size of constants exposes several inconsistencies in the code base. I've marked a few of those with `//$LIT` in the codebase. Once this change is merged into `master`, I can start picking off the `//$LIT` bugs.

Any last words before I merge into `master`?

**Comment by gbody** [comment]
Thunderbirds are GO!
